compiler_opt/rl/corpus.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""ModuleSpec definition and utility command line parsing functions."""
+
+from __future__ import annotations  # for typing .get()
+from absl import logging
+from dataclasses import dataclass
+from typing import List, Dict, Iterable, Tuple, Optional
+import abc
+import json
+import os
+import tensorflow as tf
+
+
+@dataclass(frozen=True)
+class ModuleSpec:
+  """Dataclass describing an input module and its compilation command options.
+  """
+  _exec_cmd: Tuple[str, ...]
+  _xopts: dict[str, Tuple[str, ...]]
+  name: str
+
+  def cmd(self, **kwargs) -> List[str]:
+    """Retrieves the compiler execution options,
+    optionally adding configurable, pre-set options.
+    """
+    ret_cmd: List[str] = list(self._exec_cmd)
+    for k, substitutions in kwargs.items():
+      if k in self._xopts:
+        if substitutions is not None:
+          ret_cmd += [opt.format(**substitutions) for opt in self._xopts[k]]
+      else:
+        logging.fatal(
+            'Command line addition of \'%s\' requested '
+            'but no such extra option exists.', k)
+        raise ValueError
+    return ret_cmd
+
+  @classmethod
+  def _get(cls, data_path: str, additional_flags: Tuple[str, ...],
+           delete_flags: Tuple[str, ...], xopts: Dict[str, Tuple[str, ...]]):
+    module_paths: List[str] = _load_module_paths(
+        data_path, os.path.join(data_path, 'module_paths'))
+
+    is_thinlto: bool = _has_thinlto_index(module_paths)
+    has_cmd: bool = _has_cmd(module_paths)
+
+    # TODO: (b/233935329) Per-corpus *fdo profile paths can be read into
+    # {additional|delete}_flags here
+    meta = _load_metadata(os.path.join(data_path, 'metadata.json'))
+
+    xopts.update({'output': ('-o', '{path:s}')})
+
+    module_specs: List[ModuleSpec] = []
+
+    # Future: cmd_override could have per-module overrides if needed
+    if 'global_command_override' in meta:
+      cmd_override = tuple(meta['global_command_override'])
+    else:
+      cmd_override = None
+
+    # This takes ~7s for 30k modules
+    for module_path in module_paths:
+      exec_cmd = _load_and_parse_command(
+          cmd_file=(module_path + '.cmd') if has_cmd else None,
+          ir_file=module_path + '.bc',
+          thinlto_file=(module_path + '.thinlto.bc') if is_thinlto else None,
+          additional_flags=additional_flags,
+          delete_flags=delete_flags,
+          cmd_override=cmd_override)
```

Comment:
easier to just pass thw whole metadata to _load_and_parse_command so further information can be directly extracted from meta if needed in the future